### PR TITLE
Move shape-method to Tensor-class

### PR DESCRIPTION
**Pull request by chrhansen**

### CODE REVIEW

**Comment by Arafatk** [comment]
@chrhansen Thanks.

File: lib/tensorflow/tensor.rb [hunk 1]
```diff
@@ -182,17 +182,39 @@ def ruby_array_to_c(array, type)
...
+  private
+
+  #
+  # Recursively finds the shape of the input array.
+  #
+  # * *Returns* :
+  #   - Dimension array `[[2], [4]].shape` => `[2, 1]`
+  #
+  # TODO: Handle non-rectangular arrays and raise error if mixed data types
+  def self.set_shape(value)
```
Comment (zverok) [inline lib/tensorflow/tensor.rb]: Why call it **set_**? It doesn't set's anything (not speaking of the fact that even if it had, in Ruby we never use `set_`/`get_` prefixes), just analyzes value. For me, it is something like `guess_shape` or `shape_of(value)`

Comment (chrhansen) [inline lib/tensorflow/tensor.rb]: Hi @zverok, thanks I'm familiar with the Ruby conventions and I guess it could also be called `shape_of()` instead. There's already a bare `.shape`-method (the one renamed from `.dimensions`) which is the getter for the tensor-shape. So the reason to prefix `set_`, is that if you look in the [Python-implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/ops.py#L378) the shape is inferred and _set_ when the tensor is initialized, or when someone later explicitly calls `set_shape`, and the result is held in a private variable. My idea was to follow the Tensorflow team's own design implementation and move [Line 45](https://github.com/somaticio/tensorflow.rb/blob/master/lib/tensorflow/tensor.rb#L45)

```ruby
# lib/tensorflow/tensor.rb#L45
self.shape = self.class.set_shape(value)
```
into the `set_shape` method. Thereby the shape would be _set_ only when this method was called during initialization or when someone _later_ wanted to overwrite the shape value.

Comment (zverok) [inline lib/tensorflow/tensor.rb]: I'm not sure that close following of Python conventions is the best way to design Ruby library. I've been in those discussions several times (for example, discussing some of new/existing features of [daru](https://github.com/v0dro/daru)), and whether they should look like genuine Ruby or "something familiar for similar Python library users" (pandas for daru), and the former option (make it good Ruby, not "pythonic dialect of Ruby") always wins.

In other words, if there would be "method to set shape", it should be called `shape=` in Ruby, not `set_shape` in any circumstances. (And, as far as I can see, Python's `set_shape` doesn't do analysis of incoming object's shape.)

Comment (chrhansen) [inline lib/tensorflow/tensor.rb]: @zverok My idea is to follow Ruby naming convention. I meant that the layout and design of the classes, how data is passed around, etc. should as a starting point be done similar to the Python implementation in order for the user to expect a similar behavior using both Swig wrappers and for us not to reinvent the wheel. As far as I can see, Python's `set_shape` actually does analyze the incoming shape it just happens through the [`merge_with`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/tensor_shape.py#L551) method. I think we'll be best off getting a very small subset of the API to work first than implementing all of the functionality of Tensorflow in one step 😅
